src/main/scala/org/broadinstitute/dsde/firecloud/service/PerRequest.scala

```diff
@@ -0,0 +1,117 @@
+    r.withHttpResponseHeadersMapped(h => h ++ headers).complete(response)
+    stop(self)
+  }
+
+  override val supervisorStrategy =
+    OneForOneStrategy() {
+
+      case e: RequestProcessingException =>
+        r.complete(InternalServerError, new FireCloudExceptionWithStatusCode(e.getMessage, e.getCause, e.status))
+        Stop
+      case e: Throwable =>
+        r.complete(new FireCloudExceptionWithStatusCode(e.getMessage, e.getCause, InternalServerError))
+        Stop
+    }
+}
+
+
+object PerRequest {
+
+  sealed trait PerRequestMessage
+  /**
+   * Report complete, follows same pattern as spray.routing.RequestContext.complete; examples of how to call
+   * that method should apply here too. E.g. even though this method has only one parameter, it can be called
+   * with 2 where the first is a StatusCode: RequestComplete(StatusCode.Created, response)
+   */
+  case class RequestComplete[T](response: T)(implicit val marshaller: ToResponseMarshaller[T]) extends PerRequestMessage
+
+  /**
+   * Report complete with response headers. To response with a special status code the first parameter can be a
+   * tuple where the first element is StatusCode: RequestCompleteWithHeaders((StatusCode.Created, results), header).
+   * Note that this is here so that RequestComplete above can behave like spray.routing.RequestContext.complete.
+   */
+  case class RequestCompleteWithHeaders[T](response: T, headers: HttpHeader*)(implicit val marshaller: ToResponseMarshaller[T]) extends PerRequestMessage
+
+  /** allows for pattern matching with extraction of marshaller */
+  private object RequestComplete_ {
+    def unapply[T >: Any](requestComplete: RequestComplete[T]) = Some((requestComplete.response, requestComplete.marshaller))
+  }
+
+  /** allows for pattern matching with extraction of marshaller */
+  private object RequestCompleteWithHeaders_ {
+    def unapply[T >: Any](requestComplete: RequestCompleteWithHeaders[T]) = Some((requestComplete.response, requestComplete.headers, requestComplete.marshaller))
+  }
+
+  case class WithProps(r: RequestContext, props: Props, message: AnyRef, name: String, timeout: Duration) extends PerRequest {
+    lazy val target = context.actorOf(props, name)
+  }
+}
+
+/**
+ * Provides factory methods for creating per request actors
+ */
+trait PerRequestCreator {
+  implicit def actorRefFactory: ActorRefFactory
+
+  def perRequest(r: RequestContext, props: Props, message: AnyRef, name: String = java.lang.Thread.currentThread.getStackTrace()(1).getMethodName, timeout: Duration = 1.minutes) =
+    actorRefFactory.actorOf(Props(new WithProps(r, props, message, name + System.nanoTime(), timeout)), name + System.nanoTime())
+
+  /** convenience for HttpClient */
+  def externalHttpPerRequest(r: RequestContext, request: HttpRequest) =
+    perRequest(r, Props(new HttpClient(r)), HttpClient.PerformExternalRequest(request))
```

Comment:
Yes there is a call to new in the perRequest function. good call. For the props of the "target" actor, we usually define a separate function somewhere (usually in the companion object to that actor), ala:
def queryHandlerProps = Props(classOf[QueryHandler])

Then we call perRequest like
perRequest(<requestContext>, quertyHandlerProps, <serviceMessage>)

Again, I /think/ we're safe with 'new' because we are in a trait, but the worry here is documented:
http://doc.akka.io/docs/akka/2.3.12/scala/actors.html#actors-scala
(see "Props" section)